src/server/worker/datum_iterator.go

```diff
@@ -129,16 +147,28 @@ func (d *unionDatumIterator) Next() bool {
 		d.unionIdx++
 		return d.Next()
 	}
+	d.location++
 	return true
 }
 
 func (d *unionDatumIterator) Datum() []*Input {
 	return d.iterators[d.unionIdx].Datum()
 }
 
+func (d *unionDatumIterator) DatumN(n int) []*Input {
+	if n < d.location {
+		d.Reset()
+	}
+	for d.location != n {
+		d.Next()
+	}
+	return d.Datum()
+}
```

Comment:
Same here, it doesn't seem like we should have to iterate one at a time, given that we have all the datums in memory.